src/main/java/io/jenkins/plugins/pipeline_log_fluentd_cloudwatch/LogStreamState.java

```diff
@@ -0,0 +1,505 @@
+            }
+            return client;
+        }
+
+        @Override protected String logStreamName() {
+            return logStreamNameBase + "@master";
+        }
+
+        @Override protected void ensureRunning() throws IOException {
+            boolean starting;
+            synchronized (this) {
+                starting = client == null;
+            }
+            if (starting) {
+                create(logStreamName());
+                schedule();
+            }
+        }
+
+        @Override protected synchronized void shutDown() {
+            if (client != null) {
+                client.shutdown();
+                client = null;
+            }
+        }
+
+        private void create(String logStreamName) throws IOException {
+            AWSLogs currentClient = client();
+            DescribeLogStreamsResult r = currentClient.describeLogStreams(new DescribeLogStreamsRequest(logGroupName).withLogStreamNamePrefix(logStreamName));
+            boolean found = false;
+            // TODO handle paging, in case we have a lot of similarly-named jobs
+            for (LogStream ls : r.getLogStreams()) {
+                if (ls.getLogStreamName().equals(logStreamName)) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                // First-time project.
+                LOGGER.log(Level.FINE, "Creating {0}", logStreamName);
+                currentClient.createLogStream(new CreateLogStreamRequest(logGroupName, logStreamName));
+            }
+        }
+
+        Auth authenticate() throws IOException {
+            AWSSecurityTokenServiceClientBuilder builder = AWSSecurityTokenServiceClientBuilder.standard();
+            CredentialsAwsGlobalConfiguration credentialsConfig = CredentialsAwsGlobalConfiguration.get();
+            String region = credentialsConfig.getRegion();
+            if (region != null) {
+                builder = builder.withRegion(region);
+            }
+            AmazonWebServicesCredentials jenkinsCredentials = credentialsConfig.getCredentials();
+            if (jenkinsCredentials != null) {
+                AWSStaticCredentialsProvider credentialsProvider = new AWSStaticCredentialsProvider(jenkinsCredentials.getCredentials());
+                builder.withCredentials(credentialsProvider);
+            }
+            AWSCredentialsProvider credentialsProvider = builder.getCredentials();
+            AWSCredentials masterCredentials = credentialsProvider != null ? credentialsProvider.getCredentials() : null;
+            String remotedAccessKeyId, remotedSecretAccessKey, remotedSessionToken;
+            String agentLogStreamName;
+            synchronized (agentLogStreamNames) {
+                for (int i = 1; ; i++) {
+                    String candidate = logStreamNameBase + "@agent" + i;
+                    if (agentLogStreamNames.add(candidate)) {
```

Comment:
Do we care this is unbounded? What's the likelihood of this `Set` getting a bit too large?